File: frontend/src/services/fireTrackingService.js
/**
 * Fire Tracking Service
 * Handles communication with NASA FIRMS fire detection API
 */

const API_BASE_URL = process.env.REACT_APP_API_URL || 'http://localhost:8000';

class FireTrackingService {
  constructor() {
    this.baseUrl = `${API_BASE_URL}/api/fire`;
    this.cache = new Map();
    this.loadingPromises = new Map();
    this.cacheExpiry = new Map();
    this.cacheDuration = 5 * 60 * 1000; // 5 minutes
  }

  /**
   * Fetch current fire data for Texas
   * @param {string} dataset - FIRMS dataset to use
   * @param {number} days - Number of days to fetch
   * @returns {Promise<Object>} - GeoJSON fire data
   */
  async getTexasFireData(dataset = 'VIIRS_NOAA20_NRT', days = 1) {
    const cacheKey = `fire_${dataset}_${days}`;
    
    // Return cached data if available and valid
    if (this._isCacheValid(cacheKey)) {
      console.log('🔥 Returning cached fire data');
      return this.cache.get(cacheKey);
    }

    // Return existing loading promise if already loading
    if (this.loadingPromises.has(cacheKey)) {
      return this.loadingPromises.get(cacheKey);
    }

    // Create new loading promise
    const loadingPromise = this._fetchFireData(dataset, days);
    this.loadingPromises.set(cacheKey, loadingPromise);

    try {
      const data = await loadingPromise;
      
      // Cache the result
      this.cache.set(cacheKey, data);
      this.cacheExpiry.set(cacheKey, Date.now() + this.cacheDuration);
      this.loadingPromises.delete(cacheKey);
      
      return data;
    } catch (error) {
      this.loadingPromises.delete(cacheKey);
      throw error;
    }
  }

  /**
   * Internal method to fetch fire data from API
   */
  async _fetchFireData(dataset, days) {
    try {
      const url = `${this.baseUrl}/texas?dataset=${encodeURIComponent(dataset)}&days=${days}`;
      console.log('🔥 Fetching fire data from:', url);
      
      const response = await fetch(url);
      
      if (!response.ok) {
        throw new Error(`Failed to fetch fire data: ${response.status} ${response.statusText}`);
      }

      const data = await response.json();
      
      // Validate GeoJSON structure
      if (!data || data.type !== 'FeatureCollection') {
        throw new Error('Invalid fire data format received');
      }

      console.log(`🔥 Fire data loaded: ${data.features.length} detections`);
      return data;
    } catch (error) {
      console.error('❌ Error fetching fire data:', error);
      throw new Error(`Failed to fetch fire data: ${error.message}`);
    }
  }

  /**
   * Get fire statistics for Texas
   * @param {string} dataset - FIRMS dataset to use
   * @param {number} days - Number of days for statistics
   * @returns {Promise<Object>} - Fire statistics
   */
  async getFireStatistics(dataset = 'VIIRS_NOAA20_NRT', days = 1) {
    try {
      const url = `${this.baseUrl}/statistics?dataset=${encodeURIComponent(dataset)}&days=${days}`;
      console.log('📊 Fetching fire statistics from:', url);
      
      const response = await fetch(url);
      
      if (!response.ok) {
        throw new Error(`Failed to fetch fire statistics: ${response.status} ${response.statusText}`);
      }

      const stats = await response.json();
      console.log('📊 Fire statistics loaded:', stats);
      
      return stats;
    } catch (error) {
      console.error('❌ Error fetching fire statistics:', error);
      throw new Error(`Failed to fetch fire statistics: ${error.message}`);
    }
  }

  /**
   * Get available FIRMS datasets
   * @returns {Promise<Object>} - Available datasets
   */
  async getAvailableDatasets() {
    try {
      const url = `${this.baseUrl}/datasets`;
      console.log('📋 Fetching available datasets from:', url);
      
      const response = await fetch(url);
      
      if (!response.ok) {
        throw new Error(`Failed to fetch datasets: ${response.status} ${response.statusText}`);
      }

      const datasets = await response.json();
      console.log('📋 Available datasets loaded:', datasets);
      
      return datasets;
    } catch (error) {
      console.error('❌ Error fetching available datasets:', error);
      throw new Error(`Failed to fetch available datasets: ${error.message}`);
    }
  }

  /**
   * Clear fire data cache
   */
  clearCache() {
    this.cache.clear();
    this.cacheExpiry.clear();
    this.loadingPromises.clear();
    console.log('🧹 Fire data cache cleared');
  }

  /**
   * Force refresh by clearing cache and fetching new data
   */
  async forceRefresh(dataset = 'VIIRS_NOAA20_NRT', days = 1) {
    this.clearCache();
    return await this.getTexasFireData(dataset, days);
  }

  /**
   * Clear server-side cache
   */
  async clearServerCache() {
    try {
      const url = `${this.baseUrl}/cache/clear`;
      const response = await fetch(url, { method: 'POST' });
      
      if (!response.ok) {
        throw new Error(`Failed to clear server cache: ${response.status}`);
      }

      const result = await response.json();
      console.log('🧹 Server cache cleared:', result);
      
      // Also clear local cache
      this.clearCache();
      
      return result;
    } catch (error) {
      console.error('❌ Error clearing server cache:', error);
      throw error;
    }
  }

  /**
   * Get service health status
   */
  async getServiceHealth() {
    try {
      const url = `${this.baseUrl}/health`;
      const response = await fetch(url);
      
      if (!response.ok) {
        throw new Error(`Health check failed: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error('❌ Fire service health check failed:', error);
      throw error;
    }
  }

  /**
   * Check if cached data is still valid
   */
  _isCacheValid(cacheKey) {
    if (!this.cache.has(cacheKey)) return false;
    if (!this.cacheExpiry.has(cacheKey)) return false;
    return Date.now() < this.cacheExpiry.get(cacheKey);
  }

  /**
   * Get cache statistics
   */
  getCacheStats() {
    return {
      cacheSize: this.cache.size,
      loadingPromises: this.loadingPromises.size,
      cacheDurationMs: this.cacheDuration
    };
  }

  /**
   * Format fire detection for display
   */
  formatFireDetection(feature) {
    const props = feature.properties;
    
    return {
      id: `fire_${props.latitude}_${props.longitude}_${props.acq_date}_${props.acq_time}`,
      coordinates: feature.geometry.coordinates,
      detectionTime: props.detection_time || 'Unknown',
      confidence: parseFloat(props.confidence || 0),
      confidenceLevel: props.confidence_level || 'Unknown',
      fireRadiativePower: parseFloat(props.frp || 0),
      fireIntensity: props.fire_intensity || 'Unknown',
      dataset: props.dataset_name || props.dataset || 'Unknown',
      satellite: this._extractSatellite(props.dataset || ''),
      brightness: parseFloat(props.bright_ti4 || props.brightness || props.bright_ti5 || 0),
      scan: parseFloat(props.scan || 0),
      track: parseFloat(props.track || 0)
    };
  }

  /**
   * Extract satellite name from dataset
   */
  _extractSatellite(dataset) {
    if (dataset.includes('NOAA20')) return 'NOAA-20';
    if (dataset.includes('NOAA21')) return 'NOAA-21';
    if (dataset.includes('SNPP')) return 'Suomi-NPP';
    if (dataset.includes('MODIS')) return 'MODIS';
    if (dataset.includes('LANDSAT')) return 'Landsat';
    return 'Unknown';
  }

  /**
   * Get fire marker style based on properties
   */
  getFireMarkerStyle(feature) {
    const props = feature.properties;
    const confidence = parseFloat(props.confidence || 0);
    const frp = parseFloat(props.frp || 0);
    
    // Determine color based on confidence
    let fillColor = '#ff6b6b'; // Default red
    if (confidence >= 80) {
      fillColor = '#ff3333'; // Bright red for high confidence
    } else if (confidence >= 50) {
      fillColor = '#ff6b6b'; // Medium red
    } else if (confidence >= 30) {
      fillColor = '#ff9999'; // Light red
    } else {
      fillColor = '#ffcccc'; // Very light red
    }
    
    // Determine size based on FRP (Fire Radiative Power)
    let radius = 4;
    if (frp >= 100) {
      radius = 12;
    } else if (frp >= 50) {
      radius = 10;
    } else if (frp >= 20) {
      radius = 8;
    } else if (frp >= 5) {
      radius = 6;
    }
    
    return {
      radius,
      fillColor,
      color: '#cc0000',
      weight: 2,
      opacity: 0.9,
      fillOpacity: 0.8
    };
  }
}

// Create and export singleton instance
export const fireTrackingService = new FireTrackingService();
export default fireTrackingService;

File: backend/fire_tracking_service.py
"""
Fire Tracking Service using NASA FIRMS API
Provides real-time fire detection data for Texas region
"""

import asyncio
import logging
from datetime import datetime, timedelta
from typing import Dict, List, Optional, Any
import httpx
from fastapi import HTTPException
import json
from pathlib import Path
import os
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

logger = logging.getLogger(__name__)

class FireTrackingService:
    """Service for fetching and managing fire data from NASA FIRMS API"""
    
    def __init__(self):
        # NASA FIRMS API configuration
        self.MAP_KEY = os.getenv("NASA_FIRMS_MAP_KEY", "")
        self.BASE_URL = "https://firms.modaps.eosdis.nasa.gov/api/area/csv"
        
        # Texas bounding box (West, South, East, North)
        self.TEXAS_BBOX = os.getenv("TEXAS_BBOX", "-106.65,25.84,-93.51,36.50")
        
        # Available datasets
        self.DATASETS = {
            "VIIRS_NOAA20_NRT": "VIIRS NOAA-20 Near Real-Time",
            "VIIRS_NOAA21_NRT": "VIIRS NOAA-21 Near Real-Time", 
            "VIIRS_SNPP_NRT": "VIIRS S-NPP Near Real-Time",
            "MODIS_NRT": "MODIS Near Real-Time",
            "LANDSAT_NRT": "Landsat Near Real-Time"
        }
        
        # Cache for fire data
        self.cache = {}
        self.cache_expiry = {}
        self.cache_duration = 300  # 5 minutes cache
        
    async def fetch_fire_data(self, dataset: str = "VIIRS_NOAA20_NRT", days: int = 1) -> Dict[str, Any]:
        """
        Fetch fire data from NASA FIRMS API for Texas
        
        Args:
            dataset: FIRMS dataset to use
            days: Number of days to fetch (1 = last 24h)
            
        Returns:
            GeoJSON FeatureCollection of fire detections
        """
        try:
            # Check cache first
            cache_key = f"{dataset}_{days}"
            if self._is_cache_valid(cache_key):
                logger.info(f"🔥 Returning cached fire data for {dataset}")
                return self.cache[cache_key]
            
            # Construct API URL
            url = f"{self.BASE_URL}/{self.MAP_KEY}/{dataset}/{self.TEXAS_BBOX}/{days}"
            logger.info(f"🔥 Fetching fire data from: {url}")
            
            # Make HTTP request
            async with httpx.AsyncClient(timeout=30.0) as client:
                response = await client.get(url)
                response.raise_for_status()
                
                # Parse CSV response
                csv_text = response.text
                csv_lines = csv_text.split('\n')
                logger.info(f"🔥 Received {len(csv_lines)} lines of CSV data")
                
                # Convert CSV to GeoJSON
                geojson_data = await self._csv_to_geojson(csv_text, dataset)
                
                # Cache the result
                self.cache[cache_key] = geojson_data
                self.cache_expiry[cache_key] = datetime.now() + timedelta(seconds=self.cache_duration)
                
                logger.info(f"✅ Fire data processed: {len(geojson_data['features'])} fire detections")
                return geojson_data
                
        except httpx.HTTPStatusError as e:
            logger.error(f"❌ HTTP error fetching fire data: {e}")
            if e.response.status_code == 404:
                # Return empty GeoJSON if no data available
                return self._empty_geojson()
            raise HTTPException(status_code=e.response.status_code, detail=f"FIRMS API error: {e}")
        except Exception as e:
            logger.error(f"❌ Error fetching fire data: {e}")
            raise HTTPException(status_code=500, detail=f"Fire data fetch failed: {str(e)}")
    
    async def _csv_to_geojson(self, csv_text: str, dataset: str) -> Dict[str, Any]:
        """Convert CSV response to GeoJSON format"""
        try:
            lines = csv_text.strip().split('\n')
            if len(lines) < 2:
                return self._empty_geojson()
            
            # Parse header
            headers = [h.strip() for h in lines[0].split(',')]
            logger.info(f"🔥 CSV headers: {headers}")
            
            features = []
            for line in lines[1:]:
                if not line.strip():
                    continue
                    
                values = [v.strip() for v in line.split(',')]
                if len(values) != len(headers):
                    continue
                
                # Create feature properties
                properties = {}
                for i, header in enumerate(headers):
                    if i < len(values):
                        properties[header] = values[i]
                
                # Extract coordinates
                try:
                    longitude = float(properties.get('longitude', 0))
                    latitude = float(properties.get('latitude', 0))
                    
                    # Skip invalid coordinates
                    if longitude == 0 and latitude == 0:
                        continue
                        
                    # Add additional properties for display
                    properties['dataset'] = dataset
                    properties['dataset_name'] = self.DATASETS.get(dataset, dataset)
                    properties['detection_time'] = self._format_detection_time(
                        properties.get('acq_date', ''), 
                        properties.get('acq_time', '')
                    )
                    
                    # Handle confidence - can be numeric or letter format
                    confidence_raw = properties.get('confidence', '0')
                    confidence_numeric = self._parse_confidence(confidence_raw)
                    properties['confidence'] = confidence_numeric
                    properties['confidence_level'] = self._get_confidence_level(confidence_numeric)
                    
                    # Handle FRP
                    frp_value = self._safe_float(properties.get('frp', '0'))
                    properties['frp'] = frp_value
                    properties['fire_intensity'] = self._get_fire_intensity(frp_value)
                    
                    # Create GeoJSON feature
                    feature = {
                        "type": "Feature",
                        "geometry": {
                            "type": "Point",
                            "coordinates": [longitude, latitude]
                        },
                        "properties": properties
                    }
                    features.append(feature)
                    
                except (ValueError, TypeError) as e:
                    logger.warning(f"⚠️ Skipping invalid fire detection data: {e}")
                    continue
            
            return {
                "type": "FeatureCollection",
                "features": features,
                "metadata": {
                    "dataset": dataset,
                    "dataset_name": self.DATASETS.get(dataset, dataset),
                    "total_detections": len(features),
                    "generated_at": datetime.now().isoformat(),
                    "bbox": self.TEXAS_BBOX
                }
            }
            
        except Exception as e:
            logger.error(f"❌ Error converting CSV to GeoJSON: {e}")
            return self._empty_geojson()
    
    def _format_detection_time(self, date_str: str, time_str: str) -> str:
        """Format detection date and time for display"""
        try:
            if not date_str or not time_str:
                return "Unknown"
                
            # Parse date (YYYY-MM-DD format)
            date_parts = date_str.split('-')
            if len(date_parts) != 3:
                return "Unknown"
            
            # Parse time (HHMM format)
            if len(time_str) == 4:
                hour = time_str[:2]
                minute = time_str[2:]
                return f"{date_str} {hour}:{minute} UTC"
            
            return f"{date_str} {time_str}"
        except:
            return "Unknown"
    
    def _get_confidence_level(self, confidence: float) -> str:
        """Get confidence level description"""
        if confidence >= 80:
            return "High"
        elif confidence >= 50:
            return "Medium"
        elif confidence >= 30:
            return "Low"
        else:
            return "Very Low"
    
    def _get_fire_intensity(self, frp: float) -> str:
        """Get fire intensity description based on FRP (Fire Radiative Power)"""
        if frp >= 100:
            return "Very High"
        elif frp >= 50:
            return "High"
        elif frp >= 20:
            return "Medium"
        elif frp >= 5:
            return "Low"
        else:
            return "Very Low"
    
    def _parse_confidence(self, confidence_raw: str) -> float:
        """Parse confidence value - handles both numeric and letter formats"""
        try:
            # If it's already a number, return it
            return float(confidence_raw)
        except ValueError:
            # Handle letter format (H, M, L, n)
            confidence_str = str(confidence_raw).upper().strip()
            if confidence_str == 'H':
                return 85.0  # High confidence
            elif confidence_str == 'M':
                return 65.0  # Medium confidence
            elif confidence_str == 'L':
                return 35.0  # Low confidence
            elif confidence_str == 'N':
                return 15.0  # Very low/nominal confidence
            else:
                return 50.0  # Default medium confidence
    
    def _safe_float(self, value: str) -> float:
        """Safely convert string to float"""
        try:
            return float(value)
        except (ValueError, TypeError):
            return 0.0
    
    def _empty_geojson(self) -> Dict[str, Any]:
        """Return empty GeoJSON structure"""
        return {
            "type": "FeatureCollection",
            "features": [],
            "metadata": {
                "total_detections": 0,
                "generated_at": datetime.now().isoformat(),
                "message": "No fire detections found"
            }
        }
    
    def _is_cache_valid(self, cache_key: str) -> bool:
        """Check if cached data is still valid"""
        if cache_key not in self.cache:
            return False
        if cache_key not in self.cache_expiry:
            return False
        return datetime.now() < self.cache_expiry[cache_key]
    
    async def get_fire_statistics(self, dataset: str = "VIIRS_NOAA20_NRT", days: int = 1) -> Dict[str, Any]:
        """Get fire detection statistics for Texas"""
        try:
            fire_data = await self.fetch_fire_data(dataset, days)
            features = fire_data.get('features', [])
            
            if not features:
                return {
                    "total_detections": 0,
                    "confidence_breakdown": {},
                    "intensity_breakdown": {},
                    "dataset": dataset,
                    "time_period": f"Last {days} day(s)"
                }
            
            # Analyze confidence levels
            confidence_counts = {"High": 0, "Medium": 0, "Low": 0, "Very Low": 0}
            intensity_counts = {"Very High": 0, "High": 0, "Medium": 0, "Low": 0, "Very Low": 0}
            
            for feature in features:
                props = feature.get('properties', {})
                conf_level = props.get('confidence_level', 'Unknown')
                intensity_level = props.get('fire_intensity', 'Unknown')
                
                if conf_level in confidence_counts:
                    confidence_counts[conf_level] += 1
                if intensity_level in intensity_counts:
                    intensity_counts[intensity_level] += 1
            
            return {
                "total_detections": len(features),
                "confidence_breakdown": confidence_counts,
                "intensity_breakdown": intensity_counts,
                "dataset": dataset,
                "dataset_name": self.DATASETS.get(dataset, dataset),
                "time_period": f"Last {days} day(s)",
                "last_updated": datetime.now().isoformat()
            }
            
        except Exception as e:
            logger.error(f"❌ Error getting fire statistics: {e}")
            raise HTTPException(status_code=500, detail=f"Statistics generation failed: {str(e)}")
    
    async def get_available_datasets(self) -> Dict[str, Any]:
        """Get list of available FIRMS datasets"""
        return {
            "datasets": [
                {
                    "id": dataset_id,
                    "name": dataset_name,
                    "description": f"Fire detections from {dataset_name}"
                }
                for dataset_id, dataset_name in self.DATASETS.items()
            ],
            "default_dataset": "VIIRS_NOAA20_NRT",
            "texas_bbox": self.TEXAS_BBOX
        }
    
    def clear_cache(self):
        """Clear fire data cache"""
        self.cache.clear()
        self.cache_expiry.clear()
        logger.info("🧹 Fire data cache cleared")

# Global fire tracking service instance
fire_tracking_service = FireTrackingService()
